pennylane/templates/embeddings/iqp.py

```diff
@@ -0,0 +1,249 @@
+            def circuit(features=None):
+                IQPEmbedding(features=features, wires=range(3), pattern=pattern)
+                return [qml.expval(qml.PauliZ(w)) for w in range(3)]
+
+            circuit(features=[1., 2., 3.])
+
+        Since diagonal gates commute, the order of the wire pairs has no effect on the result.
+
+        .. code-block:: python
+
+            from pennylane import numpy as np
+
+            pattern1 = [[1, 2], [0, 2], [1, 0]]
+            pattern2 = [[1, 0], [0, 2], [1, 2]]
+
+            @qml.qnode(dev)
+            def circuit(features=None, pattern=None):
+                IQPEmbedding(features=features, wires=range(3), pattern=pattern, n_repeats=3)
+                return [qml.expval(qml.PauliZ(w)) for w in range(3)]
+
+            res1 = circuit(features=[1., 2., 3.], pattern=pattern1)
+            res2 = circuit(features=[1., 2., 3.], pattern=pattern2)
+
+            assert np.allclose(res1, res2)
+
+        **Repeating the embedding**
+
+        The embedding can be repeated by specifying the ``n_repeats`` argument:
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def circuit(features=None):
+                IQPEmbedding(features=features, wires=range(3), n_repeats=4)
+                return [qml.expval(qml.PauliZ(w)) for w in range(3)]
+
+            circuit(features=[1., 2., 3.])
+
+        Every repetition uses exactly the same quantum circuit.
+
+    """
+    #############
+    # Input checks
+
+    wires = check_wires(wires)
+
+    check_no_variable(features, msg="'features' cannot be differentiable")
+
+    expected_shape = (len(wires),)
+    check_shape(
+        features,
+        expected_shape,
+        msg="'features' must be of shape {}; got {}" "".format(expected_shape, get_shape(features)),
+    )
+
+    check_type(
+        n_repeats, [int], msg="'n_repeats' must be an integer; got type {}".format(type(n_repeats))
+    )
+
+    check_type(
+        pattern,
+        [Iterable, type(None)],
+        msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+    )
+
+    if pattern is not None:
+
+        # check type
+        for p in pattern:
+            check_type(
+                p,
+                [Iterable],
+                msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+            )
+
+            for w in p:
+                check_type(
+                    w,
+                    [int],
+                    msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+                )
+
+        # check shape
+        shape = get_shape(pattern)
+        if len(shape) != 2 or shape[1] != 2:
+            raise ValueError("'pattern' must be a list of pairs of wires; got {}".format(pattern))
+
+    #####################
+
+    if pattern is None:
+        # default is an all-to-all pattern
+        pattern = [[i, j] for i in range(len(wires)) for j in range(len(wires)) if j > i]
```

Comment:
Ah, good point for future compatibility!

Reply:
Would `itertools` be a nice addition here? `itertools.combinations(wires, 2)` would give the same pattern (without the issue Tom mentioned above) although the pairs would be tuples instead of lists.

Reply:
That is very elegant solution @thisac !